Flaky Test: packetbeat/route TestDefault - `panic: runtime error: index out of range [3] with length 3`

## Flaky Test

* **Test Name:** TestDefault
* **Link:** https://github.com/elastic/beats/blob/43d5dd4342a9ff56f40fafe1ba2ad7ed0a03b6c3/packetbeat/route/route_windows_test.go#L63
* **Branch:** main
* **Artifact Link:** 
    * https://buildkite.com/elastic/beats-packetbeat/builds/5123#018fa447-ca4c-4280-bfa7-f3f4db45c455/62-229 (First failure on main)
    * https://buildkite.com/elastic/beats-packetbeat/builds/5133#018fa5bd-59d7-407f-a0d0-1680d1f84f20/108-274 (PR)
    * https://buildkite.com/elastic/beats-packetbeat/builds/5134#018fa5ca-3037-46c7-9c08-a4204b2140ae/98-265 (PR)

* **Notes:** It started failing consistently on `main` after https://github.com/elastic/beats/commit/62b4c18f2c13921814ee133f3ec4d44fe69ceba9 was merged. However this commit only removed two debug log lines from Filebeat. I cannot see any correlation between this commit and the Packetbeat test panic.

### Stack Trace

```
2024-05-23 03:18:35 EDT	=== Failed
2024-05-23 03:18:35 EDT	=== FAIL: packetbeat/route TestDefault (0.87s)
2024-05-23 03:18:35 EDT	panic: runtime error: index out of range [3] with length 3 [recovered]
2024-05-23 03:18:35 EDT		panic: runtime error: index out of range [3] with length 3
2024-05-23 03:18:35 EDT	
2024-05-23 03:18:35 EDT	goroutine 6 [running]:
2024-05-23 03:18:35 EDT	testing.tRunner.func1.2({0x899340, 0xc0000909f0})
2024-05-23 03:18:35 EDT		C:/Users/Buildkite/.go/go-1.21.10/src/testing/testing.go:1545 +0x238
2024-05-23 03:18:35 EDT	testing.tRunner.func1()
2024-05-23 03:18:35 EDT		C:/Users/Buildkite/.go/go-1.21.10/src/testing/testing.go:1548 +0x397
2024-05-23 03:18:35 EDT	panic({0x899340?, 0xc0000909f0?})
2024-05-23 03:18:35 EDT		C:/Users/Buildkite/.go/go-1.21.10/src/runtime/panic.go:914 +0x21f
2024-05-23 03:18:35 EDT	github.com/elastic/beats/v7/packetbeat/route.defaultRoute(0xc0000c2ab7?)
2024-05-23 03:18:35 EDT		C:/buildkite-agent/builds/bk-agent-prod-gcp-1716448135293881985/elastic/beats-packetbeat/packetbeat/route/route_windows_test.go:63 +0xd72
2024-05-23 03:18:35 EDT	github.com/elastic/beats/v7/packetbeat/route.TestDefault(0xc00003b520)
2024-05-23 03:18:35 EDT		C:/buildkite-agent/builds/bk-agent-prod-gcp-1716448135293881985/elastic/beats-packetbeat/packetbeat/route/route_test.go:36 +0x72
2024-05-23 03:18:35 EDT	testing.tRunner(0xc00003b520, 0x8ba630)
2024-05-23 03:18:35 EDT		C:/Users/Buildkite/.go/go-1.21.10/src/testing/testing.go:1595 +0xff
2024-05-23 03:18:35 EDT	created by testing.(*T).Run in goroutine 1
2024-05-23 03:18:35 EDT		C:/Users/Buildkite/.go/go-1.21.10/src/testing/testing.go:1648 +0x3ad
```

When I run the same tests on Linux I get an error instead of a panic:
```
tiago@millennium-falcon beats/packetbeat/route  v1.22.2  HEAD [$?] % go test ./...
--- FAIL: TestDefault (0.00s)
    route_test.go:43: unexpected error from Default(10): got:route: not found want:<nil>
    route_test.go:59: unexpected interface for family 10: got: want:wlp9s0
    route_test.go:63: unexpected interface for family 10: got:-1 want:3
FAIL
FAIL    github.com/elastic/beats/v7/packetbeat/route    0.036s
FAIL
```

I also tested it on my Windows machine and it passed, even with 100 repetitions.

--- a/CHANGELOG-developer.next.asciidoc
+++ b/CHANGELOG-developer.next.asciidoc
@@ -98,6 +98,7 @@ The list below covers the major changes between 7.0.0-rc2 and main only.
 - Fix copy arguments for strict aligned architectures. {pull}36976[36976]
 - Fix panic when more than 32767 pipeline clients are active. {issue}38197[38197] {pull}38556[38556]
 - Skip flakey metrics test on windows in filebeat httpjson input. {issue}39676[39676] {pull}39678[39678]
+- Fix flakey test on Windows 2022 in packetbeat/route. {issue}39698[39698] {pull}39822[39822]
 
 ==== Added
 

--- a/packetbeat/route/route_test.go
+++ b/packetbeat/route/route_test.go
@@ -32,7 +32,6 @@ import (
 )
 
 func TestDefault(t *testing.T) {
-	t.Skip("Flaky test: https://github.com/elastic/beats/issues/39698")
 	for _, family := range []int{syscall.AF_INET, syscall.AF_INET6} {
 		wantIface, wantIndex, wantErr := defaultRoute(family)
 		if wantErr != nil && wantErr != ErrNotFound {

--- a/packetbeat/route/route_windows_test.go
+++ b/packetbeat/route/route_windows_test.go
@@ -158,6 +158,7 @@ func fieldsN(s string, n int) []string {
 	}
 	var f []string
 	for s != "" {
+		l := len(s)
 		for i, r := range s {
 			if unicode.IsSpace(r) {
 				f = append(f, s[:i])
@@ -171,7 +172,7 @@ func fieldsN(s string, n int) []string {
 				break
 			}
 		}
-		if len(f) == n-1 {
+		if len(f) == n-1 || len(s) == l {
 			break
 		}
 	}

--- a/packetbeat/route/route_windows_test.go
+++ b/packetbeat/route/route_windows_test.go
@@ -54,6 +54,9 @@ func defaultRoute(af int) (name string, index int, err error) {
 	for inTable := false; sc.Scan(); {
 		f := strings.Fields(sc.Text())
 		if len(f) == 0 {
+			if inTable {
+				break
+			}
 			continue
 		}
 		if !inTable {
@@ -94,6 +97,9 @@ func defaultRoute(af int) (name string, index int, err error) {
 	for inTable := false; sc.Scan(); {
 		f := fieldsN(sc.Text(), 5)
 		if len(f) == 0 {
+			if inTable {
+				break
+			}
 			continue
 		}
 		if !inTable {

